src/commonMain/kotlin/fr/acinq/lightning/io/Peer.kt

```diff
@@ -287,9 +287,10 @@ class Peer(
         }
     }
 
-    fun disconnect() {
+    suspend fun disconnect() {
         if (this::socket.isInitialized) socket.close()
         _connectionState.value = Connection.CLOSED(null)
+        input.send(Disconnected)
```

Comment:
I was thinking of the situation where you might (still) have more than one channel. This is the only place where you actually choose to disconnect so I thought it would be necessary to make sure the channels were left in the 'Offline' state. 

However, if the working assumption is that the state doesn't change on disconnect, then I understand the comment in`establishConnection`.

Reply:
> I was thinking of the situation where you might (still) have more than one channel.

I don't understand why it would make anything different? If you disconnect, that disconnects all channels?

The reason we process this in `establishConnection` is because there could otherwise be a race condition, where `Disconnect` ends up being processed *after* the reconnection, which completely breaks the state machine. This happens somewhat frequently because when we're disconnected, we immediately try reconnecting.

Also, you may not be aware that you've been disconnected (if you're intermittently running in the background for example). That's ok, the only important thing to ensure is that you clean up state from a previous connection before reconnecting, which is what we currently do.

Reply:
Reverted back in [990c98e](https://github.com/ACINQ/lightning-kmp/pull/558/commits/990c98eebd8e91548ebe5493583fb8374d794e4c). I was missing the fact that  ultimately the shared connection state is held by the `socket` object in `Peer` which applies to all channels and that the `Disconnected` message is used as a prelude to a re-connection.